### ISSUE TITLE
torchtext and training of a Transformer

### ISSUE THREAD

**Issue by h56cho** [pytorch/text#806]
Hello

I am a bit confused about training my pytorch Transformer.
I am using the code below to pre-process the Penn Treebank corpus before I analyze it with my (non pre-trained) Transformer:

```python
# define the English text field
TEXT_ch2 = Field(init_token = '<sos>',
                 eos_token = '<eos>',
                 unk_token = '<unk>',
                 pad_token = '<pad>',
                 fix_length = bptt,
                 lower = True)

# split the PennTreeBank corpus into a train, val, and test set.
train_penn, val_penn, test_penn = torchtext.datasets.PennTreebank.splits(TEXT_ch2)

# build vocabulary based on the field that we just definTVD.
# (building vocabulary over all language datasets)
TEXT_ch2.build_vocab(train_penn, val_penn, test_penn,
                     specials=['<sos>','<eos>','<unk>','<pad>'])

# BPTTIterator
train_penn_iter, val_penn_iter, test_penn_iter = BPTTIterator.splits(
            (train_penn, val_penn, test_penn),
            batch_size = batch_size,
            bptt_len= bptt,
            sort_key=lambda x: len(x.text),
            sort_within_batch = True,
            shuffle = False,
            device= device,
            repeat=False)
```

My question is, since my Penn Treebank corpus are separated into train, validation, and test sets, when I train my Transformer on the Penn Treebank corpus, I would only be training the model on the Penn Treebank train and validation sets (`train_penn_iter`), am I right?

But then if I train my Transformer only on the train and validation portions of the corpus, wouldn't this mean that my Transformer will not be trained to properly handle those tokens that are contained only in the test set? so does this mean that I need to train my Transformer on the entire Penn Treebank corpus, instead of just on the training  and validation sets? But if this is the case, what then is the point of having the `split` function to separate the corpus into training, validation, and test sets? To me, this contradicts how the test sets are normally used in machine learning.

Does it make sense to "test" a Transformer on the sequences that contain the tokens which it was not trained on?

Thank you,

**Comment by zhangguanheng66** [comment]
The vocab is generated with train dataset so you will see many `<unk>` token with the test dataset. Train the model with train dataset and adjust the learnning rate with valid dataset. At the very end, you run your model with test dataset to compare the result for benchmark purpose.
